### PR TITLE
Create and modify documents

### PR DESCRIPTION
**Pull request by csansoon**
Created new `createNewDocument` and `modifyExistingDocument` methods.

Modifying a document now recomputes all documents in the commit to find changes in hash, as they might be referenced by others.

### CODE REVIEW

File: packages/core/src/data-access/documentVersions.ts [hunk 1]
```diff
@@ -3,21 +3,39 @@ import {
...
+  if (commitResult.error) return commitResult
+
+  return Result.ok(commitResult.value.mergedAt)
+}
+
 export async function getDocumentsAtCommit(
-  { commitUuid, projectId }: { commitUuid: string; projectId: number },
+  {
+    commitId,
+    commitMergedAt,
```
Comment (geclos) [inline packages/core/src/data-access/documentVersions.ts:27]: this api is unintuitive as it accepts both a commitId for a commit and another merged_at property which also belongs to commit. 

You could split it into getDocumentsAtDate and getDocumentsAtCommit methods, for example.

File: packages/core/src/data-access/documentVersions.ts [hunk 2]
```diff
@@ -64,44 +82,39 @@ export async function getDocumentsAtCommit(
...
-      documentsAtDraft.find((d2) => d2.documentUuid !== d.documentUuid),
+    .filter(
+      (d) =>
+        documentsAtDraft.find((d2) => d2.documentUuid === d.documentUuid) ===
+        undefined,
     )
     .concat(documentsAtDraft)
 
   return Result.ok(totalDocuments)
 }
```
Comment (geclos) [inline packages/core/src/data-access/documentVersions.ts:108]: hmm hadn't seen this method before. It's a clever query! You should be able to improve code by splitting in smaller chunks. You really only need three operations for getting documents at a particular commit afaict:

- fetch main branch documents
- fetch commit documents
- merge main branch and draft documents

Can we split this method in these 3 concepts?

File: packages/core/src/services/documents/create.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    commitId,
+    commitMergedAt: commitResult.value.mergedAt,
+  })
+  if (currentDocuments.error) return currentDocuments
+
+  if (
+    existsAnotherDocumentWithSamePath({
+      documents: currentDocuments.value,
+      path,
+    })
```
Comment (geclos) [inline packages/core/src/services/documents/create.ts:32]: hmm we are still doing validations in services.. drizzle is really lacking compared to active record

File: packages/core/src/services/documents/create.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  ) {
+    return Result.error(
+      new LatitudeError('A document with the same path already exists'),
```
Comment (geclos) [inline packages/core/src/services/documents/create.ts:35]: use a badrequest error instead

File: packages/core/src/services/documents/modify.test.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    }).then((r) => r.unwrap())
+
+    expect(changedDocuments.length).toBe(2)
+    expect(
+      changedDocuments.find((d) => d.path === 'referenced/doc'),
+    ).toBeDefined()
+    expect(changedDocuments.find((d) => d.path === 'unmodified')).toBeDefined()
+  })
+
+  it('renaming a document creates a change to all other documents that reference it', async (ctx) => {
```
Comment (geclos) [inline packages/core/src/services/documents/modify.test.ts:97]: also to all its children given you have to change the paths

File: packages/core/src/services/documents/modify.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  path?: string
+  content?: string | null
+  deletedAt?: Date | null
+}) {
+  const commitResult = await getEditableCommit(commitId)
+  if (commitResult.error) return commitResult
+
+  const currentDocuments = await getDocumentsAtCommit({
+    commitId,
+    commitMergedAt: commitResult.value.mergedAt,
```
Comment (geclos) [inline packages/core/src/services/documents/modify.ts:60]: hmm yeah it's not clear what this will do

File: packages/core/src/services/documents/modify.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  })
+  if (currentDocuments.error) return currentDocuments
```
Comment (geclos) [inline packages/core/src/services/documents/modify.ts:62]: hmmm we this result error check in a million places already I wonder if we can improve. How about a try catch for this whole method and you just wrap whatever error you get in a Result.error?

File: packages/core/src/services/documents/modify.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            set: newDocumentVersion,
+          })
+          .returning()
+          .then((r) => r[0]!)
+      }),
+    )
+
+    return Result.ok(results.find((r) => r.documentUuid === documentUuid)!)
+  })
+}
```
Comment (geclos) [inline packages/core/src/services/documents/modify.ts:136]: hard to follow what is happening in this method can we split it in smaller chunks? Can we do it with less filters, destructings and finds and less if else statements?

File: packages/core/src/services/documents/utils.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  findCommitById,
+  Result,
+  TypedResult,
+} from '@latitude-data/core'
+import { ForbiddenError, LatitudeError } from '$core/lib/errors'
+
+export async function getEditableCommit(
+  commitId: number,
+): Promise<TypedResult<Commit, LatitudeError>> {
+  const commit = await findCommitById({ id: commitId })
```
Comment (geclos) [inline packages/core/src/services/documents/utils.ts:13]: all other finds return a result this one should too

File: packages/core/src/data-access/documentVersions.ts (hunk 1 above)
Comment (csansoon) [reply packages/core/src/data-access/documentVersions.ts:27]: I only added `mergedAt` for performance, as it is needed for retrieving the documents, and several parts of the code that use this function already have this information so I didn't want to fetch this if it's not necessary. But I agree that it's unintuitive.

File: packages/core/src/services/documents/modify.test.ts (hunk 5 above)
Comment (csansoon) [reply packages/core/src/services/documents/modify.test.ts:97]: This is just testing a change of the path of a single document. This does not affect any other document.

The action of renaming a folder will be done as a different service, where it will rename all paths with the same folder.

File: packages/core/src/services/documents/create.ts (hunk 3 above)
Comment (andresgutgon) [reply packages/core/src/services/documents/create.ts:34]: I always thought model validation level was not the right approach. In the end you usually have one or two services that change the entity. I think is better to have in those services the validation.

File: apps/web/src/actions/documents/create.ts [hunk 10]
```diff
@@ -15,6 +15,13 @@ export const createDocumentVersionAction = withProject
...
   .handler(async ({ input }) => {
-    const result = await createDocumentVersion(input)
+    const commit = await findCommitByUuid({
+      projectId: input.projectId,
+      uuid: input.commitUuid,
+    }).then((r) => r.unwrap())
+    const result = await createNewDocument({
+      commitId: commit.id,
+      path: input.path,
+    })
```
Comment (andresgutgon) [inline apps/web/src/actions/documents/create.ts:25]: The finding and saving should be all inside `createNewDocument`. The less code we put inside actions the better

File: apps/web/src/components/Sidebar/index.tsx [hunk 11]
```diff
@@ -9,9 +9,9 @@ export default async function Sidebar({
   commitUuid: string
   projectId: number
 }) {
+  const commit = await findCommitByUuid({ projectId, uuid: commitUuid })
```
Comment (andresgutgon) [inline apps/web/src/components/Sidebar/index.tsx:12]: I feel this `findCommit` it's going to be used on several components on the server at the same time. https://github.com/latitude-dev/latitude-llm/blob/main/apps/web/src/app/(private)/_data-access/index.ts#L11C32-L11C37

File: packages/core/src/data-access/documentVersions/getDocumentByPath.ts [hunk 12]
```diff
@@ -16,7 +17,9 @@ export async function getDocumentByPath(
   db = database,
 ) {
   try {
-    const result = await getDocumentsAtCommit({ projectId, commitUuid }, db)
+    const commit = await findCommitByUuid({ projectId, uuid: commitUuid }, db)
+    if (commit.error) return commit
+    const result = await getDocumentsAtCommit({ commitId: commit.value.id }, db)
```
Comment (andresgutgon) [inline packages/core/src/data-access/documentVersions/getDocumentByPath.ts:22]: 👾

File: packages/core/src/data-access/documentVersions/getDocumentsAtCommit.test.ts [hunk 13]
```diff
@@ -1,8 +1,10 @@
 import { HEAD_COMMIT } from '$core/constants'
+import { modifyExistingDocument } from '$core/services'
```
Comment (andresgutgon) [inline packages/core/src/data-access/documentVersions/getDocumentsAtCommit.test.ts:2]: Why not a more standard name. `updateDocument`

File: packages/core/src/services/documents/modify.test.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
+import { listCommitChanges } from '$core/data-access'
+import useTestDatabase from '$core/tests/useTestDatabase'
+import { describe, expect, it } from 'vitest'
+
+import { mergeCommit } from '../commits/merge'
+import { modifyExistingDocument } from './modify'
+
+useTestDatabase()
```
Comment (andresgutgon) [inline packages/core/src/services/documents/modify.test.ts:8]: Why this?

File: packages/core/src/services/documents/modify.test.ts [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+describe('modifyExistingDocument', () => {
```
Comment (andresgutgon) [inline packages/core/src/services/documents/modify.test.ts:10]: DItto use more standard `update`

File: packages/core/src/data-access/documentVersions/getDocumentByPath.ts (hunk 12 above)
Comment (csansoon) [reply packages/core/src/data-access/documentVersions/getDocumentByPath.ts:22]: ?

**Review by andresgutgon — APPROVED** [review]

File: packages/core/src/services/documents/create.ts (hunk 3 above)
Comment (geclos) [reply packages/core/src/services/documents/create.ts:34]: Validations should be closest to the source of truth as possible so to make it difficult to bypass them, this means validations in db when possible, validations closest to db when db validations aren't possible. Services are two steps removed from the db, model just one. Compose your model with validations and you'll be safer.

File: packages/core/src/data-access/documentVersions/getDocumentsAtCommit.ts [hunk 16]
```diff
@@ -61,33 +60,60 @@ export async function getDocumentsAtCommit(
       ),
     )
 
-  const documentsAtPreviousMergedCommits =
-    documentsAtPreviousMergedCommitsResult.map((d) => d.document_versions)
+  return documentsFromMergedCommits
+}
```
Comment (geclos) [inline packages/core/src/data-access/documentVersions/getDocumentsAtCommit.ts:64]: fantastic improvement 👌🏼 !
